Add libvips 8.8.1

jcupitt opened the pull request:
Hello @conda-forge/help-c-cpp , here's a recipe for libvips that works, but would benefit from some feedback. 

## Background

[libvips](https://libvips.github.io/libvips/) is an open-source image processing library. It's [fast and needs only a small amount of memory](https://github.com/libvips/libvips/wiki/Speed-and-memory-use). It's popular for web servers, for imaging, especially large images, and in bioinformatics.

It has a Python binding in pip, [pyvips](https://pypi.org/project/pyvips/) and a [proposed conda recipe](https://github.com/jcupitt/staged-recipes/tree/add-pyvips-2.1.8), but before we can add that to conda, we obviously need a recipe for the underlying C/C++ library.

There have been a few previous attempts to make a recipe for libvips: this PR is supposed to draw on that experience. Issue https://github.com/libvips/libvips/issues/673 has the related discussion and links.

## What works

This recipe seems to build OK on my laptop. It also builds with CI, ie. this:

```
john@kiwi:~/GIT/conda/staged-recipes (master)$ ./.circleci/run_docker_build.sh
```

runs successfully.

## Questions

It has a number of issues though:

- `openslide` support is missing, since that's in bioconda and this is a conda-forge recipe. How should this be resolved? Should we try to move openslide to conda-forge?
- I was unable to find conda packages for `lcms` and `libexif`.  These are relatively important, I suppose we should package these too.
- Relatively minor missing libs: `libgsf-1` (image pyramid support), `niftiio` (medical images), `openexr` (movie business images), `libheif` (needed for HEIC support). It would be good to find packages for these too, but not essential.
- `orc` is missing, so there's no SIMD support. Things like convolution will be up to about 3x slower.
- I noticed `imagemagick` is built with `jpeg`, not `libjpeg-turbo`, so we must use `jpeg` too. It would be good if IM could move to turbo as well. It's a trivial change with a 3x speedup. Would this be possible?
- I had to pin `pango` to 1.40.14 to get `librsvg` and `imagemagick` to go on together. Again, if `librsvg` and `imagemagick` could move to a common and recent `pango` library, that would be very helpfui.

## Other platforms

I've only tested on linux, but I've enabled macOS as well, as it should work. It should build on win, but we'd need to steal a `build.bat` from somewhere, if I understand the process.

## Checklist

- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vend other packages
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your
      recipe (see
      [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos)
      for more details)
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/libvips```) and found some lint.

Here's what I've got...


For **recipes/libvips**:

 * There are 1 too many lines.  There should be one empty line at the end of the file.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libvips```) and found it was in an excellent condition.

hmaarrfk commented on recipes/libvips/README.md:
```diff
@@ -0,0 +1,160 @@
+# -*- mode: jinja -*-
+-->
+
+About libvips
```
I think this is going to get clobbered. Consider making it more succinct and adding it to the recipe.

hmaarrfk commented on recipes/libvips/build.sh:
```diff
@@ -0,0 +1,12 @@
+export CPPFLAGS="-I${PREFIX}/include ${CPPFLAGS}"
+export LDFLAGS="-L${PREFIX}/lib ${LDFLAGS}"
+export LD_LIBRARY_PATH="${PREFIX}/lib:${LD_LIBRARY_PATH}"
+export CFLAGS="-O3 ${CFLAGS}"
```
I don't think you should do `-O3`. that is a conda-forge wide choice.

jcupitt replied:
My reasoning was that libvips is marked up for the gcc autovectoriser, and that's enabled at O3 or above.  It gives a significant boost over O2 for some operations.

I'll remove it.

hmaarrfk replied:
is there a way to add just the one flag you want? and a comment to explain?

hmaarrfk commented on recipes/libvips/build.sh:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+export CPPFLAGS="-I${PREFIX}/include ${CPPFLAGS}"
```
Try commenting this out. It should just work.

hmaarrfk commented on recipes/libvips/build.sh:
```diff
@@ -0,0 +1,12 @@
+export CXXFLAGS="-O3 ${CXXFLAGS}"
+
+./configure --prefix="${PREFIX}"
+( make > make.out ) || ( cat make.out && exit 1 )
```
make.out won't really be stored anywhere permanent, maybe just replace with this `make`

hmaarrfk commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,121 @@
+    - libmatio
+    - libpng
+    - librsvg 
+    - libtiff
```
Inspire yourself from https://github.com/conda-forge/opencv-feedstock/blob/master/recipe/meta.yaml#L49

to find where different dependencies should be included.

jcupitt replied:
Oh that's really helpful, thanks!

hmaarrfk commented on COPYING:
```diff
@@ -0,0 +1,502 @@
+                  GNU LESSER GENERAL PUBLIC LICENSE
```
I think you can remvoe this whole file, and simply add a single like to point to the `COPYING` file in the recipe.

hmaarrfk commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,121 @@
+  home: https://libvips.github.io/libvips
+  license: LGPL-2.1
+  license_family: LGPL
+  license_file: COPYING
```
Basically, this line will look at the top level of the source, and find that the COPYING file exists.

hmaarrfk commented:
I've tried to get libvips working before. Had to resort to Ubuntu's package manager. This is awesome.

Hopefully the cleanup I suggested will aid in a more thorough review. But as it stands, it seems pretty straight forward and you should be able to do quite a bit of cleanup.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libvips```) and found it was in an excellent condition.


I do have some suggestions for making it better though...


For **recipes/libvips**:

 * license_file entry is missing, but is expected.

jcupitt commented:
Hey Mark, this is great! Thank you very much for the careful review. I added the licence_file to stop the linter complaining, but you seem to be saying we should use the one from the tarball, is that correct? I've probably misunderstood.

I've left my "to - do" list as a comment in the `meta.yml`.

I suppose we leave the unfilled dependencies for now and hope to update them later.

hmaarrfk commented on recipes/libvips/build.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+./configure --prefix="${PREFIX}" \
+	&& make \
```
i dont think the && style is needed. conda will run thi script and return on the first faile command

hmaarrfk replied:
> -ftree-vectorize and by default at -O3. To allow vectorization on powerpc* platforms also use -maltivec. On i?86 and x86_64 platforms use -msse/-msse2. To enable vectorization of floating point reductions use -ffast-math or -fassociative-math.

i thinik we already add the sse, but you can add the other flags.

Does llvm work too? or just gcc

hmaarrfk replied:
`-mtune=haswell -ftree-vectorize`
`-march=nocona`

Are already in in the CFLAGS. `nocona implies -msse3`
https://wiki.gentoo.org/wiki/GCC_optimization

You are missing the math flags, which presumably you need?

jcupitt replied:
OK, if the vectoriser is on it should be fine. 

I'll remove the `&&`. I guess conda runs this file with `set -e` and then sourcing it, is that right? So I guess I can remove the `#!/bin/bash` too.

jcupitt replied:
I usually use `-mnative`, but I assumed conda set something like that already.

hmaarrfk replied:
I forget what mnative does. Unfortunately you don't always control the exact compilation machine, so it is better to explicit about the exact processor you are targeting.

I mean, I'm not sure if the flags will be accepted, but there is a much higher chance of them getting accepted than '-03'

hmaarrfk replied:
I think conda will assume it is bash unless otherwise specified. I mean that shebang line isn't super distracting but the bash one liner isn't too friendly.

jcupitt replied:
Dockerfiles have made me very used to huge `x && y && z` lines, it's the standard idiom heh.

I've removed it.

hmaarrfk commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,131 @@
+about:
+  home: https://libvips.github.io/libvips
+  license: LGPL-2.1
+  license_family: LGPL
```
```suggestion
  license_family: LGPL
  license_file: COPYING
```

hmaarrfk replied:
You still need the license file line, but it will get it from the source.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/libvips```) and found it was in an excellent condition.

hmaarrfk commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,132 @@
+    - glib
+    - hdf5
+    - imagemagick
+    - jpeg
```
The only thing I can add is that conda forge has been adding run exports to many of these libraries so technically you only have to include them in the gostsection

hmaarrfk replied:
The issue, is that the lintner won't help you here, so you have to go through them manually. This is probably a good feature request for the linter

hmaarrfk commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,132 @@
+
+build:
+  number: 0
+  skip: true  # [win]
```
This is a c library correct? That would be added a build time dependency? If so, could you add a run exports?

jcupitt replied:
The core is a mix of C and C++, but with a strict C-only API. There are then separate C and C++ APIs built on top of that. 

Could you point me to the docs or an example for run exports?

hmaarrfk replied:
Docs are a little confusing:
https://github.com/conda-forge/libpng-feedstock/blob/master/recipe/meta.yaml#L18

With that line, you only need to include it as a build time dependency, conda-forge will understand that "build time" means that you need the associated "runtime pinning", and will make the package incompatible when the minor version changes. Patch version changes are OK.

https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements

hmaarrfk commented:
You can likely ask the boys to rerender those recepies that you had to pin so that their pinnings move up

hmaarrfk commented:
Yeah, i know. Honestly, I hate having to maintain so many different files,
so I'll oten fold in easy recipes like this one into a single meta file.

On Thu, Jul 25, 2019 at 9:04 AM John Cupitt <notifications@github.com>
wrote:

> *@jcupitt* commented on this pull request.
> ------------------------------
>
> In recipes/libvips/build.sh
> <https://github.com/conda-forge/staged-recipes/pull/8912#discussion_r307284382>
> :
>
> > @@ -0,0 +1,6 @@
> +#!/usr/bin/env bash
> +
> +./configure --prefix="${PREFIX}" \
> +	&& make \
>
> Dockerfiles have made me very used to huge x && y && z lines, it's the
> standard idiom heh.
>
> I've removed it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/conda-forge/staged-recipes/pull/8912?email_source=notifications&email_token=AAAV7GF3WSGCCF6KEXZIXC3QBGQHPA5CNFSM4IGPVTGKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB7SCM6Y#discussion_r307284382>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAV7GCFJPUVUUL34DSYSBTQBGQHPANCNFSM4IGPVTGA>
> .
>

jcupitt commented:
I had a go at adding a `run_exports` line, hopefully I've not messed up.

I removed all the `run:` lines, then added them back until I had no warnings. I think this means I removed `run:` lines for packages with a `run_exports` (is that correct?).

hmaarrfk commented on recipes/libvips/meta.yaml:
```diff
@@ -74,27 +77,13 @@ requirements:
     - imagemagick
-    - jpeg
+    - gdk-pixbuf
     - libmatio
```
nice, the scary part is that these run_exports aren't always well defined. but nice cleanup.

hmaarrfk commented:
> I think this means I removed run: lines for packages with a run_exports (is that correct?).

Yes

jcupitt commented:
OK, then we're good to merge, from my POV.

hmaarrfk commented:
Ok, now that linux and mac are done, how hard is windows to compile?

jcupitt commented:
Pretty easy -- it's regularly built with MSVC for things like the node binding. I usually cross-compile from linux. But it would need someone other than me to make and test a `build.bat` I think.

jcupitt commented:
Actually I should qualify that -- the basic DLLs are always cross-compiled from linux (since the build system is autotools), just the C++ API is rebuilt with MSVC to match node.

I would leave win for now, unless you also use a cross-compiler.

hmaarrfk commented:
Yeah. Conda forge compiles natively. If you need support for mingw I think you can do that too.

If you use standard cmake tools, the open CV recipe might be of help.

I can't really pull, but if you think this will pass mustard without the windows build, then ok. Generally I think there needs to be a good argument, but I've used the "will work on soon" for pytorch myself.

jcupitt commented:
Yes, if conda has mingw / msys it should be simple for someone to add later. I don't have easy access to a win machine, so it'd be awkward for me to develop.

marcelotrevisani commented on recipes/libvips/build.sh:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+./configure --prefix="${PREFIX}" 
+make 
```
Is it possible to compile it with multiple cores?
```suggestion
make -j${CPU_COUNT}
```

jcupitt commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,121 @@
+    - libtiff
+    - libwebp
+    - pango =1.40.14
+    - poppler
```
Hello @scopatz, yes, imagemagick and librsvg both have complicated pango pins and libvips needs to be able to link to both of them. This pango is (I think?) the only choice until IM and librsvg can agree on a more recent pango.

hmaarrfk replied:
we might need to issue a rerender to the dependencies you need.

hmaarrfk replied:
I just removed the pin and was able to build successfully
```
    imagemagick:           7.0.8_54-pl526h39023e4_0      conda-forge
    pango:                 1.42.4-he7ab937_0          conda-forge
    librsvg:               2.44.14-h11c8777_0            conda-forge
```

Are these the versions you were hoping for?

hmaarrfk replied:
I think the libpng pinning was recently loosened, so i think it **was** problematic before.

jcupitt replied:
That's great news! I'll try removing the pin.

jcupitt commented on recipes/libvips/build.sh:
```diff
@@ -0,0 +1,12 @@
+
+export CPPFLAGS="-I${PREFIX}/include ${CPPFLAGS}"
+export LDFLAGS="-L${PREFIX}/lib ${LDFLAGS}"
+export LD_LIBRARY_PATH="${PREFIX}/lib:${LD_LIBRARY_PATH}"
```
I think you might be looking at an outdated version of this branch -- this was changed a while ago.

scopatz replied:
yup

jcupitt commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,121 @@
+    - pkg-config
+    - pango =1.40.14
+    - poppler
+    - zlib
```
I think that's an outdated list -- the current version just has cc, c++, make and pkg-config.

jcupitt commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,121 @@
+    - libmatio
+    - librsvg 
+    - libwebp
+    - pango =1.40.14
```
I think so, see above.

hmaarrfk commented:
Such a strange failure now.

jcupitt commented:
It's failing on the new libgsf dependency. Perhaps it needs some more time to propagate?

I'll try triggering a build again in 24h.

jcupitt commented:
It seems to be all working.

Remaining issues:

- `openslide` support is missing, since that's in bioconda and this is a conda-forge recipe. How should this be resolved? Should we try to move `openslide` to conda-forge?
- Relatively minor missing libs: `niftiio` (medical images), `openexr` (movie business images), `libheif` (needed for HEIC support). It would be good to find packages for these too, but not essential.
- `orc` is missing, so there's no SIMD support. Things like convolution will be up to about 3x slower.
- I noticed `imagemagick` is built with `jpeg`, not `libjpeg-turbo`, so we must use `jpeg` too. It would be good if IM could move to turbo as well. It's a trivial change with a 3x speedup for jpg load/save. Would this be possible?

hmaarrfk commented:
All these seem like minor improvement that can likely be iterated on when your recipe gets approved. I would get it to a state you are happy enough with, and push for the approval process

hmaarrfk commented:
There has been some interest about moving to libjpeg turbo. I think with the migrators it might be possible to do it. again, I don't see this as a blocker for this recepi.

jcupitt commented:
I'd like to get orc added still, but I agree the rest can wait. I'll ping this again when orc is in.

hmaarrfk commented:
@jcupitt you will be able to get orc added after merging. The idea at this stage is mostly to make sure that you get things like pinning correct, c-flags, and dependencies. Others can always help in your feedstock when you ping them, you just won't be at their mercy to make changes. Just my two cents.

jcupitt commented:
OK, I think this is ready to merge!

- All the important deps are now in conda-forge
- I've read the build logs and they are all being picked up correctly
- The compiler flags being used look good
- It seems to build correctly on all platforms except win (various packages that libvips needs indirectly, like gdk-pixbuf, seem not to be packaged for win)

I'll ask gitter.

chrisburr commented on recipes/libvips/meta.yaml:
```diff
@@ -0,0 +1,103 @@
+
+extra:
+  recipe-maintainers:
+    - sebastian-luna-valero
```
@sebastian-luna-valero @strets123 Are you happy to be listed as maintainers of this recipe?

jcupitt replied:
They volunteered in this thread: https://github.com/libvips/libvips/issues/673 though it wouldn't hurt to ask again.

sebastian-luna-valero replied:
Yes, happy to keep involved. Thanks everyone!

jcupitt replied:
Looking at stret123's profile page, he's been quiet for two weeks, I guess he's on holiday. I'll remove his name for now, and perhaps replace if it he asks. There will be a 8.8.2 rather soon.

chrisburr replied:
That's good enough for me, I had only checked the participants on this page.

bgruening commented:
> openslide support is missing, since that's in bioconda and this is a conda-forge recipe. How should this be resolved? Should we try to move openslide to conda-forge?

Yes please move it from bioconda to conda-forge. Can you please also remove then the recipe from bioconda.

Thanks!

jcupitt commented:
Hi @bgruening -- already done! I moved it yesterday.

Add to conda-forge: https://github.com/conda-forge/staged-recipes/pull/9169

Remove from bioconda: https://github.com/bioconda/bioconda-recipes/pull/16938

From my to-do list above, the remaining items are:


- Relatively minor missing libs: `niftiio` (medical images), `openexr` (movie business images), `libheif` (needed for HEIC support). It would be good to find packages for these too, but not essential.
- I noticed `imagemagick` is built with `jpeg`, not `libjpeg-turbo`, so we must use `jpeg` too. It would be good if IM could move to turbo as well. It's a trivial change with a 3x speedup for jpg load/save. Would this be possible?

I'll look into switching to libjpeg-turbo separately. Perhaps something with a migration? I don't really understand the conda tooling around this.

The other missing deps are minor and can go in later.

bgruening commented:
@jcupitt awesome. Thanks a lot. I agree those steps can be added in the feedstock. Rock on!

bgruening commented:
> I'll look into switching to libjpeg-turbo separately. Perhaps something with a migration? I don't really understand the conda tooling around this.

This will be a global conda-forge effort. You can drive this or wait until a automatic PR pops up in your feedstock :)

jcupitt commented:
I asked on gitter re. jpeg -> libjpeg-turbo and opinion seemed mostly positive, but there were concerns regarding library version compat on macos. It seems like a relatively small effort and a big win to me, so I'm keen to try to push it.

I'll have a quick look at that once libvips is in.

jcupitt commented:
OK, all maintainers have said "yes", I think we can merge.

jcupitt commented:
Thank you @chrisburr !
